### DISCUSSION TITLE
When end id is reached can we break the execution?

### DISCUSSION THREAD

**Discussion by kaushalruparel** [aahnik/tgcf#310] [Q&A] [answered]
When we have provided end id in config, can we break the execution to immediately come out. For source with huge number of files/ids (85k+) tgcf takes around 10-15 mins to come out and also gets gethistoryrequest flood wait sometimes. Can we do a break instead of continue?
code snippet for ref:

`if forward.end and last_id > forward.end:
                    continue`

Also below condition needs to be rewritten
`if forward.end and last_id > forward.end:`
to 
`if forward.end and last_id >= forward.end:`
to stop exactly at end id. Currently it forwards end id + 1 message

**Comment by tissole** [comment, marked as answer]
I run a test with an 80k msg channel. I forwarded 8 msgs using `offset:` and `end:` setting. The script does not hang. So maybe this happens only when you forward a large number of msgs? How many have you tried to forward?

Have you tried to use a [delay](https://github.com/aahnik/tgcf/wiki/How-to-configure-tgcf-%3F)? Maybe it will solve this flood error.

>Currently it forwards end id + 1 message 

> Also below condition needs to be rewritten
if forward.end and last_id > forward.end:
to
if forward.end and last_id >= forward.end:
to stop exactly at end id. 

Yes, I already describe this error in #198. Thanks for the solution, now is working as expected!

**Reply by kaushalruparel** [reply]
I forwarded 10 messages and after it forwarded end id, it waited for around 10 mins before coming out and giving back the control on prompt. Then I forwarded 1k messages and here also it took 10 mins before giving back control after reaching end id and this time I get flood control warning also on gethistoryrequest. 

In both the cases I had a delay set of 2.12 seconds.

**Reply by kaushalruparel** [reply]
And I forgot to mention that I tested with break and it comes out immediately on the next iteration of loop where `last_id >= forward.end:
`
Although I see no error or unexpected behaviour with break, I am not sure if this is the proper way of coming out of the program

Below is what I tested with and it comes out of inner and outer loops immediately on hitting this condition as TRUE
```
            if forward.end and last_id >= forward.end:
                    break
```

**Reply by tissole** [reply]
I was able to replicate this. It happens when the msgs are at the beginning or in the middle of a big channel. First I tried with msgs from the near end of the chat, no hanging because there are few msgs till the end. If there are a lot of msgs till the channel's end the script seems to iterate through all of them and this is the reason for the flood errors. TG has a limit of 2000 requests per hour and after this is starting to show these flood error msgs.

When I used `break` the script stops immediately after the last msg is forwarded. So probably this has to be done from now on. You could do a pull request.

**Reply by kaushalruparel** [reply]
Thank You, will initiate a pull request
